src/VisualStudio/Core/Def/LanguageService/AbstractCreateServicesOnTextViewConnection.cs

```diff
@@ -45,56 +47,54 @@ protected virtual Task InitializeServiceForOpenedDocumentAsync(Document document
         {
             if (!_initialized)
             {
-                var token = _listener.BeginAsyncOperation(nameof(InitializeServicesAsync));
-                InitializeServicesAsync().CompletesAsyncOperation(token);
-
                 _initialized = true;
+                // use `null` to trigger per VS session intialization task
+                _workQueue.AddWork((DocumentId?)null);
             }
         }
 
         void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
         {
         }
 
-        private void InitializeServiceOnDocumentOpened(object sender, DocumentEventArgs e)
+        private async ValueTask ProcessBatchDocumentOpenedAsync(ImmutableSegmentedList<DocumentId?> documentIds, CancellationToken cancellationToken)
         {
-            if (e.Document.Project.Language != _languageName)
+            foreach (var documentId in documentIds)
             {
-                return;
-            }
-
-            var token = _listener.BeginAsyncOperation(nameof(InitializeServiceForOpenedDocumentOnBackgroundAsync));
-            InitializeServiceForOpenedDocumentOnBackgroundAsync(e.Document).CompletesAsyncOperation(token);
-
-            async Task InitializeServiceForOpenedDocumentOnBackgroundAsync(Document document)
-            {
-                await TaskScheduler.Default;
-
-                // Preload project completion providers on a background thread since loading extensions can be slow
-                // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1488945
-                if (document.GetLanguageService<CompletionService>() is not null)
-                    _ = CompletionService.GetProjectCompletionProviders(document.Project);
-
-                await InitializeServiceForOpenedDocumentAsync(document).ConfigureAwait(false);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (documentId is null)
+                {
+                    InitializePerVSSessionServices();
+                }
+                else if (Workspace.CurrentSolution.GetDocument(documentId) is Document document && document.Project.Language == _languageName)
+                {
+                    await InitializeServiceForOpenedDocumentAsync(document).ConfigureAwait(false);
+                }
             }
         }
 
-        private async Task InitializeServicesAsync()
-        {
-            await TaskScheduler.Default;
+        private void QueueWorkOnDocumentOpened(object sender, DocumentEventArgs e)
+            => _workQueue.AddWork(e.Document.Id);
 
+        private void InitializePerVSSessionServices()
+        {
             var languageServices = Workspace.Services.GetExtendedLanguageServices(_languageName);
 
             _ = languageServices.GetService<ISnippetInfoService>();
```

Comment:
Should the other GetLazyImportedProviders also be removed here?

Reply:
Those are MEF imported providers and still need to be pre-loaded here (i.e. once per VS session , when a document is opened)